cmd/nvidia-cdi-hook/cudacompat/cudacompat_test.go

```
@@ -0,0 +1,131 @@
+# limitations under the License.
+*/
+
+package cudacompat
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	testlog "github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCompatLibs(t *testing.T) {
+	logger, _ := testlog.NewNullLogger()
+
+	testCases := []struct {
+		description            string
+		contents               map[string]string
+		options                options
+		expectedCompatContents string
+	}{
+		{
+			description: "empty root",
+			options: options{
+				hostDriverVersion: "222.55.66",
+			},
+		},
+		{
+			description: "compat lib is newer; no ldcache",
+			contents: map[string]string{
+				"/usr/local/cuda/compat/libcuda.so.333.88.99": "",
+			},
+			options: options{
+				hostDriverVersion: "222.55.66",
+			},
+		},
+		{
+			description: "compat lib is newer; ldcache",
+			contents: map[string]string{
+				"/etc/ld.so.cache":                            "",
+				"/etc/ld.so.conf.d/.hidden":                   "",
+				"/usr/local/cuda/compat/libcuda.so.333.88.99": "",
+			},
+			options: options{
+				hostDriverVersion: "222.55.66",
+			},
+			expectedCompatContents: "/usr/local/cuda/compat\n",
+		},
+		{
+			description: "compat lib is older; ldcache",
+			contents: map[string]string{
+				"/etc/ld.so.cache":                            "",
+				"/etc/ld.so.conf.d/.hidden":                   "",
+				"/usr/local/cuda/compat/libcuda.so.111.88.99": "",
+			},
+			options: options{
+				hostDriverVersion: "222.55.66",
+			},
+			expectedCompatContents: "",
+		},
+		{
+			description: "compat lib has same major version; ldcache",
+			contents: map[string]string{
+				"/etc/ld.so.cache":                            "",
+				"/etc/ld.so.conf.d/.hidden":                   "",
+				"/usr/local/cuda/compat/libcuda.so.222.88.99": "",
+			},
+			options: options{
+				hostDriverVersion: "222.55.66",
+			},
+			expectedCompatContents: "",
+		},
+		{
+			description: "driver version empty; ldcache",
```

Question -- I am just trying to grok this scenario. IIUC this means we always use the compat libraries if we cannot discover the driver version of the libraries present on the host. Does libnvidia-container have the same behavior? Taking a step back, shouldn't failing to detect the driver version from the host not result in an error?

---

Yes, failing to detect the driver version will cause a failure in `libnvidia-container`. That behaviour is not captured here since the driver version extraction happens earlier in the process.

The reason that I didn't want to raise an error in this case is that I didn't want to cause a container to fail to start due to an error in the runtime -- more drastically changing the behaviour. I think I can change the behaviour of the hook to be a no-op if the driver version is not specified though.

---

I have change the behaviour to a no-op if the driver root is empty.